src/background/redirect-protection.js

```diff
@@ -0,0 +1,230 @@
+    return undefined;
+  }
+
+  if (
+    Object.keys(options.redirectProtection.disabled).some((domain) =>
+      hostname.endsWith(domain),
+    )
+  ) {
+    return undefined;
+  }
+
+  return REDIRECT_PROTECTION_PAGE_URL + '?url=' + btoa(url);
+}
+
+if (__PLATFORM__ !== 'firefox') {
+  chrome.webNavigation.onBeforeNavigate.addListener(
+    (details) => {
+      if (
+        details.frameId === 0 &&
+        details.url &&
+        !details.url.startsWith(REDIRECT_PROTECTION_PAGE_URL)
+      ) {
+        redirectUrlMap.set(details.tabId, details.url);
+      }
+    },
+    {
+      url: [{ schemes: ['http', 'https'] }],
+    },
+  );
+
+  chrome.tabs.onRemoved.addListener((tabId) => {
+    redirectUrlMap.delete(tabId);
+
+    chrome.declarativeNetRequest
+      .updateSessionRules({
+        removeRuleIds: [REDIRECT_PROTECTION_SESSION_ID_RANGE.start + tabId],
+      })
+      .catch(() => {});
+  });
+
+  OptionsObserver.addListener(
+    async function redirectProtectionExceptions(options, lastOptions) {
+      if (options.redirectProtection.enabled) {
+        const disabledDomains = Object.keys(
+          options.redirectProtection.disabled,
+        );
+        const lastDisabledDomains = lastOptions
+          ? Object.keys(lastOptions.redirectProtection.disabled)
+          : [];
+
+        const hasChanges =
+          !lastOptions ||
+          !lastOptions.redirectProtection.enabled ||
+          JSON.stringify(disabledDomains.sort()) !==
+            JSON.stringify(lastDisabledDomains.sort());
+
+        if (hasChanges) {
+          try {
+            const removeRuleIds = await getDynamicRulesIds(
+              REDIRECT_PROTECTION_ID_RANGE,
+            );
+
+            const addRules = createRedirectProtectionExceptionRules(
+              disabledDomains,
+              REDIRECT_PROTECTION_ID_RANGE.start,
+            );
+
+            await chrome.declarativeNetRequest.updateDynamicRules({
+              removeRuleIds,
+              addRules,
+            });
+          } catch (e) {
+            console.error(
+              '[redirect-protection] Error updating exception rules:',
+              e,
+            );
+          }
+        }
+      } else if (lastOptions?.redirectProtection.enabled) {
+        const removeRuleIds = await getDynamicRulesIds(
+          REDIRECT_PROTECTION_ID_RANGE,
+        );
+        if (removeRuleIds.length) {
+          await chrome.declarativeNetRequest.updateDynamicRules({
+            removeRuleIds,
+          });
+        }
+      }
+    },
+  );
+}
+
+chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
+  if (message.action === 'getRedirectUrl') {
+    if (__PLATFORM__ === 'firefox') {
+      return false;
+    }
+    const url =
+      sender.tab && sender.tab.id
+        ? redirectUrlMap.get(sender.tab.id) || null
+        : null;
+    sendResponse({ url });
+    return false;
+  }
+
+  if (message.action === 'allowRedirect') {
```

Comment:
Allow and disallow redirects are still side effects to the user action - if something goes wrong, the proper state is never re-created.

In our architecture, we should re-create the state as a listener to option changes. We can still look if the rules are there, and don't remove all and set all (we can be smart with that).

Reply:
those are session rules, meant for a very short time. We dont want to put the on Options. If anything, we should rather remove them ASAP, not sure if webNavigation as on Firefox will work as expected.